### PR TITLE
Support all images in EC2 Instance Image Filters

### PR DESCRIPTION
**Pull request by Sutto**
This is an attempted solution at #979.

TL;DR: This changes the internals of ResourceQuery to allow "escaping" the Enum spec and adjusts how the instance
image filter works to take advantage of this. No more `Could not locate image for instance` simply because it's a public image.

## General Train of Thought

1. `DescribeImages` defaults to returning _all_ accessible images (ouch).
2. The current AMI resource thus scopes to `Owners: ["Self"]` for just our images (logical, right?)
3. the `ec2` instance image filter uses the current AMI resource, looking up via ID.
4. ... This is limited to our resources.
5. What if we call `get_resources` to fetch specific ids?
6. `enum_spec` specifies the filter, which is merged in on the `get_resources` call via the `ResourceQuery`.
7. Oh, what if we work around that?
8. ... and call get_resources
9. We need to avoid the cache!
10. ... but we'll needlessly invoke the API then.
11. Smartly batch them up and then save them on the resource.
12. Intelligently fetch them!

This uses the `c7n:key` style augment like in some of the account checks, to avoid needlessly refetching data.
We don't cache the AMI results on the AMI level, but this ins't an issue because the AMI resource is only for _our images_.

## Use Cases

* Find using public / market place images (we should use an internal hardened image)
* Using the microsoft / windows VMs

## TODO:

* [ ] Tests - May need to run in the custodian skunkworks?
* [x] How can we avoid the `skip_extra_args` argument? Maybe pull out to a higher level?

### CODE REVIEW

**Comment by kapilt** [comment]
Wow ;-) this was a rabbit hole. So a few comments and a question. Nutshell, the root of the issue seems to be enum_spec extra args is used unconditionally.

Checking around to see what uses the extra args

```
from c7n.manager import resources
from c7n.resources import load_resources

load_resources()

for name, klass in resources.items():
    spec = getattr(klass.resource_type, 'enum_spec', None)
    if spec and spec[-1]:
        print name, spec
```

shows only two.

```
ami ('describe_images', 'Images', {'Owners': ['self']})
ebs-snapshot ('describe_snapshots', 'Snapshots', {'OwnerIds': ['self']})
```

i think in both of those cases if we're querying by id, we should drop the extra args.

i'm not as concerned with the cache aspects here, when calling the resource manager to fetch by id,  we can explicitly say to avoid the cache. the resource annotations are local to a policy, they are not cached across policies, so they would only be a net win if you have multiple image filters in the same policy, which seems like a minority use case, and also given this is a bulk api, we're still incurring just a single api call per image filter.  do you have multiple uses of this filter on the same policy where the api call avoidance would be meaningful?

one alternative impl route, is if instead of handling the owners self via extra args we do it as a query filter with default, which we do for a few other resources, and let the caller effectively specify overrides when they instantiate the resource manager. this would allow us to drop the extra args stuff entirely.

side note.. on the topic of resource annotations,  one note of caution we're not yet stable on our annotation keys, not all are properly prefixed, and there' still some contention about the prefix, `c7n.` causes issues for some indexers, and `c7n:` is problematic for jmespath queries on annotations for the report subcommand (via --field). i was thinking of switching to `c7n-`.

**Comment by kapilt** [comment]
nevermind the side comment on annotation, jmespath does support a literal query to via quoting to escape the colon.

**Comment by Sutto** [comment]
True, the caching on resources isn't actually much of an optimization - but it does mean atleast we can preload the image info a batch (vs. loading on an individual api) - so not much need. The actualy cache cache, we avoid in this call for that reason :). That said, multiple filters per policy would be useful (don't have a specific example) for finding `or`'s between conditions - e.g. owned by us OR in owned by X, things like that? Definitely useful in terms of the per-policy stuff.

I think the c7n- is a great idea - and possibly standardising on naming along the lines of "c7n-x-y" (vs. c7n-XY).

Do you have a specific example of the query filter w/ a default? That approach would be cleaner - We need to make sure we scope the get resources call (even with specified ids), since otherwise DescribeImages is *huge* (So it's not just about the extra arguments being an issue, it's also a "index the resources and fetch by id" vs. "invoke the api lookup with the given ids").

(I hope this wasn't too ranty, I was jumping back and forth through paragraphs as I was writing it).

**Comment by Sutto** [comment]
@kapilt this is ready for a re-review - I've significantly simplified the scope of the PR, and made it work differently.

I haven't removed the extra_args item yet, because it's still used in emr.py - but there is a bug in that as well that needs to be looked at. I'll add an issue to remove it if that's something you think is worth doing going forward.
